plugins/lines/prism-lines.js

```diff
@@ -0,0 +1,130 @@
+(function () {
+
+	if (typeof self === 'undefined' || !self.Prism || !self.document) {
+		return;
+	}
+
+	/**
+	 * Returns style declarations for the element
+	 * @param {Element} element
+	 */
+	Prism.hooks.add('before-insert', function (env) {
+
+		// works only for <code> wrapped inside <pre> (not inline)
+		var pre = env.element.parentNode;
+		if (!pre || !/pre/i.test(pre.nodeName)) {
+			return;
+		}
+
+		// we are not welcome here :(
+		if (pre.className.includes('nolines')) {
+			return;
+		}
+
+		// support for line-numbers behaviour
+		var dataStart = 1;
+		if (pre.hasAttribute('data-start')) {
+			dataStart = parseInt(pre.getAttribute('data-start'), 10);
+			if (isNaN(dataStart)) {
+				dataStart = 1;
+			}
+			pre.style.setProperty('--counter-start', dataStart - 1);
+                }
+
+		function createLine(lineNumber, prefix) {
+			var line = document.createElement('div');
+			var anchor = document.createElement('a');
+			if (prefix) {
+				var target = prefix + '.' + lineNumber;
+				line.id = target;
+				anchor.href = '#' + target;
+			}
+			line.setAttribute('class', 'line');
+			line.appendChild(anchor);
+			return line;
+		}
+
+		function splitNode(node) {
+			if (node.hasChildNodes()) {
```

Comment:
I am not sure that I understand your comments for this particular place, since this function is the working horse of the entire plugin and that check is deciding whether we dive into recursion (an element has children with the newline embedded somewhere) or we just do a simple split of a text node.

The warning over there is to catch an unexpected case of getting something else than a text node, with no children, yet containing a newline inside. I fail to imagine such a case, but if it happens there will be a warning.

I run this code against almost 150 MB of different sources highlighted with Prism, not a single warning..

Reply:
The `node` variable might be an element (e.g. `<span>`) that just happens not to have elements. You haven't run into any problems because we explicitly test all language definitions so that they don't generate empty tokens. So Prism basically guarantees that `<span>`s won't be empty, **but other plugins don't**.
We still have to be able to handle this case.

Reply:
The `splitNode()` is invoked _only_ if there is a new line character detected in the node under the scope, and that character at that time is in a text node only, so there is no chance that when `splitNode()` runs it will get an empty element with no children, yet that contains a new line character, so that conditional branch is good as it is, I believe, and no additional handling is required.